File: code/algo/getoffer/c++/2/bst_flatten.cpp
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
  TreeNode *increasingBST(TreeNode *root) {
    TreeNode *dummyNode = new TreeNode(0);
    TreeNode *resNode = dummyNode;

    inorder(root, resNode);

    return dummyNode->right;
  }

private:
  void inorder(TreeNode *node, TreeNode *&resNode) {
    if (!node) {
      return;
    }
    inorder(node->left, resNode);

    // Modify the node pointers during in-order traversal
    resNode->right = node;
    node->left = nullptr;
    resNode = node;

    inorder(node->right, resNode);
  }
};

File: code/algo/getoffer/go/2/bst_flatten.go
package main

func increasingBST(root *TreeNode) *TreeNode {
	dummyNode := &TreeNode{}
	resNode := dummyNode

	var inorder func(*TreeNode)
	inorder = func(node *TreeNode) {
		if node == nil {
			return
		}
		inorder(node.Left)

		// 在中序遍历的过程中修改节点指向
		resNode.Right = node
		node.Left = nil
		resNode = node

		inorder(node.Right)
	}
	inorder(root)

	return dummyNode.Right
}
